src/csharp/Grpc.Core/CallCredentials.cs

```diff
@@ -48,81 +47,43 @@ public static CallCredentials Compose(params CallCredentials[] credentials)
         /// <param name="interceptor">authentication interceptor</param>
         public static CallCredentials FromInterceptor(AsyncAuthInterceptor interceptor)
         {
-            return new MetadataCredentials(interceptor);
+            return new AsyncAuthInterceptorCredentialsConfiguration(interceptor);
         }
 
         /// <summary>
-        /// Creates native object for the credentials.
+        /// Populates this call credential instances.
+        /// You never need to invoke this, part of internal implementation.
         /// </summary>
-        /// <returns>The native credentials.</returns>
-        internal abstract CallCredentialsSafeHandle ToNativeCredentials();
-    }
+        public abstract void InternalPopulateConfiguration(CallCredentialsConfiguratorBase configurator, object state);
 
-    /// <summary>
-    /// Client-side credentials that delegate metadata based auth to an interceptor.
-    /// The interceptor is automatically invoked for each remote call that uses <c>MetadataCredentials.</c>
-    /// </summary>
-    internal sealed class MetadataCredentials : CallCredentials
-    {
-        readonly AsyncAuthInterceptor interceptor;
-
-        /// <summary>
-        /// Initializes a new instance of <c>MetadataCredentials</c> class.
-        /// </summary>
-        /// <param name="interceptor">authentication interceptor</param>
-        public MetadataCredentials(AsyncAuthInterceptor interceptor)
-        {
-            this.interceptor = GrpcPreconditions.CheckNotNull(interceptor);
-        }
-
-        internal override CallCredentialsSafeHandle ToNativeCredentials()
+        private class CompositeCallCredentialsConfiguration : CallCredentials
```

Comment:
nit: can we rename `CompositeCallCredentialsConfiguration` to just `CompositeCallCredentials`? Because it inherits from `CompositeCallCredentials`, changing the suffix harms readability IMO.

Same for `AsyncAuthInterceptorCredentialsConfiguration`

Reply:
Done.